packages/squiggle-lang/src/rescript/Reducer/Reducer_Dispatch/Reducer_Dispatch_BuiltInMacros.res

```diff
@@ -3,120 +3,136 @@
 
 type expression = ExpressionT.expression
-
-type reducerFn = (
-  expression,
-  ExpressionT.bindings,
-) => result<ExpressionValue.expressionValue, errorValue>
+type environment = ExpressionValue.environment
+type errorValue = Reducer_ErrorValue.errorValue
 
 let dispatchMacroCall = (
-  list: list<expression>,
+  macroExpression: expression,
   bindings: ExpressionT.bindings,
-  reduceExpression: reducerFn,
-): result<expression, 'e> => {
-  let rec replaceSymbols = (expression: expression, bindings: ExpressionT.bindings): result<
-    expression,
-    errorValue,
-  > =>
-    switch expression {
-    | ExpressionT.EValue(EvSymbol(aSymbol)) =>
-      switch bindings->Belt.Map.String.get(aSymbol) {
-      | Some(boundExpression) => boundExpression->Ok
-      | None => RESymbolNotFound(aSymbol)->Error
-      }
-    | ExpressionT.EValue(_) => expression->Ok
-    | ExpressionT.EBindings(_) => expression->Ok
-    | ExpressionT.EList(list) => {
-        let racc = list->Belt.List.reduceReverse(Ok(list{}), (racc, each: expression) =>
-          racc->Result.flatMap(acc => {
-            each
-            ->replaceSymbols(bindings)
-            ->Result.flatMap(newNode => {
-              acc->Belt.List.add(newNode)->Ok
-            })
-          })
-        )
-        racc->Result.map(acc => acc->ExpressionT.EList)
+  environment,
+  reduceExpression: ExpressionT.reducerFn,
+): result<expression, errorValue> => {
+  let doBindStatement = (bindingExpr: expression, statement: expression, environment) =>
+    switch statement {
+    | ExpressionT.EList(list{ExpressionT.EValue(EvCall("$let")), symbolExpr, statement}) => {
+        let rExternalBindingsValue = reduceExpression(bindingExpr, bindings, environment)
+
+        rExternalBindingsValue->Result.flatMap(externalBindingsValue => {
+          let newBindings = Bindings.fromValue(externalBindingsValue)
+          let rNewStatement = Bindings.replaceSymbols(newBindings, statement)
+          rNewStatement->Result.map(newStatement =>
+            eFunction(
+              "$setBindings",
+              list{newBindings->Bindings.toExternalBindings->eRecord, symbolExpr, newStatement},
+            )
+          )
+        })
       }
+    | _ => REAssignmentExpected->Error
     }
 
-  let doBindStatement = (statement: expression, bindings: ExpressionT.bindings) => {
+  let doBindExpression = (bindingExpr: expression, statement: expression, environment) =>
     switch statement {
-    | ExpressionT.EList(list{
-        ExpressionT.EValue(EvCall("$let")),
-        ExpressionT.EValue(EvSymbol(aSymbol)),
-        expressionToReduce,
-      }) => {
-        let rNewExpressionToReduce = replaceSymbols(expressionToReduce, bindings)
+    | ExpressionT.EList(list{ExpressionT.EValue(EvCall("$let")), symbolExpr, statement}) => {
+        let rExternalBindingsValue = reduceExpression(
+          bindingExpr,
+          Belt.Map.String.fromArray([("x", ExpressionValue.EvNumber(666.))]),
```

Comment:
What's happening here? Is this a mistake?

Reply:
No. 
The code is handling valid functions. However, as Sam is finding out it is not good at cases with errors.
I have put that number there intentionally. If that number leaks out to the results than that means there is an unhandled error case.